src/contracts/SavingCircles.sol

```diff
@@ -0,0 +1,276 @@
+    if (
+      !allowedTokens[_circle.token] || _circle.depositInterval == 0 || _circle.depositAmount == 0
+        || _circle.maxDeposits == 0 || _circle.circleStart == 0 || _circle.currentIndex != 0
+        || _circle.owner == address(0) || _circle.members.length < MINIMUM_MEMBERS
+    ) revert InvalidCircle();
+
+    for (uint256 i = 0; i < _circle.members.length; i++) {
+      address _member = _circle.members[i];
+      if (_member == address(0)) revert InvalidCircle();
+      isMember[_id][_member] = true;
+    }
+    circles[_id] = _circle;
+
+    emit CircleCreated(
+      _id, _circle.name, _circle.members, _circle.token, _circle.depositAmount, _circle.depositInterval
+    );
+  }
+
+  /**
+   * @notice Make a deposit into a specified circle
+   * @param _id Identifier of the circle
+   * @param _value Amount of the token to deposit
+   */
+  function deposit(bytes32 _id, uint256 _value) external override nonReentrant {
+    _deposit(_id, _value, msg.sender);
+  }
+
+  /**
+   * @notice Make a deposit on behalf of another member
+   * @param _id Identifier of the circle
+   * @param _value Amount of the token to deposit
+   * @param _member Address to make a deposit for
+   */
+  function depositFor(bytes32 _id, uint256 _value, address _member) external override nonReentrant {
+    _deposit(_id, _value, _member);
+  }
+
+  /**
+   * @notice Make a withdrawal from a specified circle
+   * @param _id Identifier of the circle
+   */
+  function withdraw(bytes32 _id) external override nonReentrant {
+    Circle storage _circle = circles[_id];
+
+    if (!_withdrawable(_id)) revert NotWithdrawable();
+    if (_circle.members[_circle.currentIndex] != msg.sender) revert NotWithdrawable();
+    if (_circle.currentIndex >= _circle.maxDeposits) revert NotWithdrawable();
+
+    uint256 _withdrawAmount = _circle.depositAmount * (_circle.members.length);
+
+    for (uint256 i = 0; i < _circle.members.length; i++) {
+      balances[_id][_circle.members[i]] = 0;
+    }
+
+    _circle.currentIndex = (_circle.currentIndex + 1) % _circle.members.length;
+    bool success = IERC20(_circle.token).transfer(msg.sender, _withdrawAmount);
+    if (!success) revert TransferFailed();
+
+    emit FundsWithdrawn(_id, msg.sender, _withdrawAmount);
+  }
+
+  /**
+   * @notice Set if a token can be used for saving circles
+   * @param _token Token to update the status of
+   * @param _allowed Can be used for saving circles
+   */
+  function setTokenAllowed(address _token, bool _allowed) external override onlyOwner {
+    allowedTokens[_token] = _allowed;
+
+    emit TokenAllowed(_token, _allowed);
+  }
+
+  /**
+   * @notice Decommission an existing saving circle
+   * @dev Returns all deposits to members
+   * @param _id Identifier of the circle
+   */
+  function decommission(bytes32 _id) external override {
+    Circle storage _circle = circles[_id];
+
+    if (_circle.owner != msg.sender) {
+      if (!isMember[_id][msg.sender]) revert NotMember();
```

Comment:
what's the value of having access control here for the circle members as opposed to permissionless? 

My concern is that practically what will happen is we /chain coop will have to decommission these, not the members themselves...

Reply:
i usually like to avoid "godmode" functions, but if we think we need it we can add it. it just gives the circle members complete control which is a good or bad thing depending on perspective